src/util/schnorr.rs

```diff
@@ -76,9 +80,77 @@ impl TweakedPublicKey {
     pub fn new(key: PublicKey) -> TweakedPublicKey {
         TweakedPublicKey(key)
     }
-    
-    /// Returns the underlying public key 
+
+    /// Returns the underlying public key
     pub fn into_inner(self) -> PublicKey {
         self.0
     }
+}
+
+/// A BIP340-341 serialized schnorr signature with the corresponding hash type.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct SchnorrSig {
+    /// The underlying schnorr signature
+    pub sig: secp256k1::schnorrsig::Signature,
+    /// The corresponding hash type
+    pub hash_ty: SchnorrSigHashType,
+}
+
+impl SchnorrSig {
+
+    /// Deserialize from slice
+    pub fn from_slice(sl: &[u8]) -> Result<Self, SchnorrSigError> {
+        if sl.len() == SCHNORRSIG_SIGNATURE_SIZE {
+            // default type
+            let sig = secp256k1::schnorrsig::Signature::from_slice(sl)?;
+            Ok( SchnorrSig { sig, hash_ty : SchnorrSigHashType::Default })
+        } else {
+            match sl.split_last() {
+                Some((hash_ty, sig)) => {
+                    let hash_ty = SchnorrSigHashType::from_u8(*hash_ty)
+                        .map_err(|_| SchnorrSigError::InvalidSighashType(*hash_ty))?;
+                    let sig = secp256k1::schnorrsig::Signature::from_slice(sig)?;
+                    Ok(SchnorrSig { sig, hash_ty })
+                },
+                None => {
+                    Err(SchnorrSigError::Secp256k1(secp256k1::Error::InvalidSignature))?
+                },
```

Comment:
This could be written as an `ok_or_else` after the `split_last`. That would get rid of the explicit pattern match.

Reply:
Also if this is kept, use `return Err(...)` instead of `Err(...)?`